core/dataset_operations.cpp

```diff
@@ -106,7 +106,13 @@ DataArray concatenate(const DataConstProxy &a, const DataConstProxy &b,
 
 Dataset concatenate(const DatasetConstProxy &a, const DatasetConstProxy &b,
                     const Dim dim) {
-  Dataset result;
+  const auto da = Dimensions(a.dimensions());
+  const auto db = Dimensions(b.dimensions());
+  Dataset result(std::map<std::string, Variable>(),
+                 concat(a.coords(), b.coords(), dim, da, db),
+                 concat(a.labels(), b.labels(), dim, da, db),
+                 concat(a.masks(), b.masks(), dim, da, db),
+                 std::map<std::string, Variable>());
```

Comment:
I think I see why you added the constructor, but it has to be solved differently. Looking at `concat` it appears that it is not making any use of the features that require ordering. Can you template the `dimsA` and `dimsB` arguments in `concat` so it also accepts other types (such as `unordered_map`) that have `operator[]`?